File: src/main/java/com/jore/epoc/bo/orders/IncreaseQualityOrder.java
package com.jore.epoc.bo.orders;

import org.hibernate.annotations.CompositeType;

import com.jore.datatypes.money.Money;
import com.jore.epoc.bo.accounting.FinancialAccounting;
import com.jore.epoc.bo.message.MessageLevel;

import jakarta.persistence.AttributeOverride;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;

@Entity
public class IncreaseQualityOrder extends AbstractSimulationOrder {
    @AttributeOverride(name = "amount", column = @Column(name = "increase_quality_amount"))
    @AttributeOverride(name = "currency", column = @Column(name = "increase_quality_currency"))
    @CompositeType(com.jore.datatypes.hibernate.MoneyCompositeUserType.class)
    private Money increaseQualityAmount;

    @Override
    public void execute() {
        if (getCompany().getAccounting().checkFunds(increaseQualityAmount, getExecutionMonth().atEndOfMonth())) {
            increaseQuality();
            book(getExecutionMonth().atDay(1), "Increase quality", FinancialAccounting.SERVICES, FinancialAccounting.BANK, increaseQualityAmount);
            addMessage(MessageLevel.INFORMATION, "IncreaseQuality", getExecutionMonth(), increaseQualityAmount);
            setExecuted(true);
        } else {
            addMessage(MessageLevel.WARNING, "NoIncreaseInQualityDueToFunding", getExecutionMonth(), increaseQualityAmount, getCompany().getAccounting().getBankBalance(getExecutionMonth().atEndOfMonth()));
        }
    }

    @Override
    public Money getAmount() {
        return increaseQualityAmount;
    }

    @Override
    public int getSortOrder() {
        return 7;
    }

    @Override
    public String getType() {
        return "Increase quality";
    }

    public void setAmount(Money increaseQualityAmount) {
        this.increaseQualityAmount = increaseQualityAmount;
    }

    private void increaseQuality() {
        Money pricePerPercentPoint = getCompany().getSimulation().getSettings().getPricePerPercentPointQuality();
        double increase = increaseQualityAmount.divide(pricePerPercentPoint).doubleValue();
        increase /= 100;
        getCompany().setQualityFactor(getCompany().getQualityFactor() + increase);
    }
}

File: EpocServer/Sources/App/Models/Orders/IncreaseQualityOrder.swift
import Foundation

public final class IncreaseQualityOrder: SimulationOrder {
    public var id: Int64?
    public var executionMonth: YearMonth = .of(2020, 1)
    public var isExecuted: Bool = false
    public weak var company: Company!
    public var increaseQualityAmount: Money = Money.of("CHF", 0)

    public var sortOrder: Int { 7 }
    public var type: String { "Increase quality" }

    public func getAmount() -> Money { increaseQualityAmount }

    public func execute() {
        if company.accounting.checkFunds(increaseQualityAmount, valueDate: executionMonth.atEndOfMonth()) {
            increaseQuality()
            book(bookingDate: executionMonth.atDay(1), bookingText: "Increase quality",
                 debitAccount: FinancialAccounting.SERVICES, creditAccount: FinancialAccounting.BANK, amount: increaseQualityAmount)
            addMessage(.information, "IncreaseQuality", executionMonth, increaseQualityAmount)
            isExecuted = true
        } else {
            addMessage(.warning, "NoIncreaseInQualityDueToFunding", executionMonth, increaseQualityAmount, company.accounting.getBankBalance(executionMonth.atEndOfMonth()))
        }
    }

    private func increaseQuality() {
        let pricePerPercentPoint = company.simulation.settings!.getPricePerPercentPointQuality()
        var increase = increaseQualityAmount.divide(pricePerPercentPoint).doubleValue
        increase /= 100
        company.qualityFactor += increase
    }
}
